packages/next/types/index.d.ts

```diff
@@ -78,13 +80,15 @@ export type GetStaticPaths = () => Promise<{
   fallback: boolean
 }>
 
-export type GetServerSideProps = (context: {
+export type GetServerSideProps<
+  P extends { [key: string]: any } = { [key: string]: any }
+> = (context: {
   req: IncomingMessage
   res: ServerResponse
   params?: ParsedUrlQuery
   query: ParsedUrlQuery
   preview?: boolean
   previewData?: any
-}) => Promise<{ [key: string]: any }>
+}) => Promise<P>
```

Comment:
Wait, this looks incorrect. It should be:
```suggestion
}) => Promise<{ props: P }>
```